### ISSUE TITLE
Replace use of `Typeable` by `ShowProxy`

### ISSUE THREAD

**Issue by edsko** [input-output-hk/ouroboros-network#2432]

**Comment by mrBliss** [comment]
We have `Typeable` constraints for two reasons:

1. `Exception`s. But we have since hidden the type arguments (e.g., make them existential) to make classification of exceptions in the `ErrorPolicy` easier. So this reason no longer applies. If an exception still mentions type arguments, it should be changed.
2. Deriving via for `NoUnexpectedThunks`. Instead, either use the variant that take a manually specified name (e.g., [`UseIsNormalFormNamed`](https://github.com/input-output-hk/cardano-prelude/blob/71ea865408f2e03e6d6832359423546699730849/src/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs#L339) vs. `UseIsNormalForm`) or introduce a new newtype with a `ShowProxy` constraint (instead of a `Typeable`) on its `NoUnexpectedThunks` instance and use that with deriving via.

**Comment by kderme** [comment]
This sometimes leads to Network code, ie 
Ouroboros.Consensus.Node.run -> Ouroboros.Consensus.Network.NodeToClient.mkApps ->  Ouroboros.Network.Driver.runPeer

I was wondering if I should be touching this code or there is a similar ticket for the network side.

**Comment by mrBliss** [comment]
The network team were the ones that insisted on introducing `ShowProxy`, which is redundant w.r.t. `Typeable`, hence this ticket. If it's just a matter of removing the `Typeable` constraint in network code, go ahead, we'll ask for a review from @coot when the PR is ready. If removing the `Typeable` constraint in the network code leads to issues, we can ask @coot for help.

**Comment by kderme** [comment]
> Exceptions. But we have since hidden the type arguments (e.g., make them existential) to make classification of exceptions in the ErrorPolicy easier. So this reason no longer applies. If an exception still mentions type arguments, it should be changed.

`data DecoderFailure failure` from network code has, so I'm trying to do the same.

**Comment by kderme** [comment]
Also currently the default implementation for `showProxy` uses `Typeable`. I think we can use Generic instead for this. For example for instances like this
```
instance ShowProxy blk => ShowProxy (SerialisedHeader blk) where
    showProxy _ = "SerialisedHeader " ++ showProxy (Proxy :: Proxy blk)
```
can instead get a default instance by following the K1 type constructor https://hackage.haskell.org/package/base-4.14.0.0/docs/GHC-Generics.html#t:K1 (and ignoring the D1, C1, constructors). This will help with many other types, which currently need a Typeable instance for their type parameters:

```
instance (Typeable c, Typeable ext, Typeable ext')
    => ShowProxy (SimpleBlock' c ext ext') where

instance (Typeable c, Typeable ext, Typeable ext')
    => ShowProxy (Header (SimpleBlock' c ext ext')) where

instance (Typeable c, Typeable ext)
    => ShowProxy (GenTx (SimpleBlock c ext)) where

instance Typeable xs => ShowProxy (HardForkBlock xs) where\
instance Typeable xs => ShowProxy (Header (HardForkBlock xs)) where
instance Typeable xs => ShowProxy (HardForkApplyTxErr xs) where
instance Typeable xs => ShowProxy (GenTx (HardForkBlock xs)) where
instance Typeable xs => ShowProxy (TxId (GenTx (HardForkBlock xs))) where
instance Typeable xs => ShowProxy (Query (HardForkBlock xs)) where
```
and many more and remove furthermore Typeable constraints.

That's something I can do as part of this issue if we think it's worth it.

**Comment by mrBliss** [comment]
Closed by #2538.
